Feature test for Wikipedia

danielamormocea opened the pull request:
Fixes #841
Firefox Voice is now remembering context for the wikipedia site. If it's alright, I'll add the other baseUrls.

ianb commented on extension/intents/navigation/navigation.js:
```diff
@@ -51,19 +52,31 @@ intentRunner.registerIntent({
+
+    if (service === undefined) {
+      service = await serviceList.detectServiceFromActiveTab(metadata.search);
+      tab = await context.activeTab();
```
Note this code currently fails if you search on a site that isn't supported. This is what I get:

```
Error in intent navigation.bangSearch : TypeError: can't access property "toLowerCase", name is null ddgBangServiceName@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/background/serviceList.js:17:16
ddgBangSearchUrl@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/searching.js:56:28
run@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/intents/navigation/navigation.js:63:35
async*runIntent@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/background/intentRunner.js:278:18
runUtterance@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/background/intentRunner.js:255:10
@moz-extension://52ad6e95-d3ed-b243-adfd-fce5310dae53/background/main.js:40:25
```

Probably instead of giving an error this should open up a regular search with `site:${(new URL(tab.url)).origin}` added to the query.

ianb commented on extension/intents/navigation/navigation.js:
```diff
@@ -51,19 +52,31 @@ intentRunner.registerIntent({
+      browser.runtime.sendMessage({
+        type: "closePopup",
+        sender: "find",
+      });
```
Be sure to run `npm run format`, it'll fix small indentation problems like this

ianb commented on extension/intents/navigation/navigation.toml:
```diff
@@ -5,6 +5,7 @@ match = """
   images (of | for) [query] [service=images]
   (do a |) (search | search on | query on | lookup on | look up on | look on | look in | look up in | lookup in) (my |) [service:serviceName] (for | for the |) [query] (for me |)
   (do a |) (search | query ) my [service:serviceName] (for | for the |) [query] (for me|)
+  (do a |) (search | query ) (this site | here) (for | for the |) [query] (for me|)
```
It would be good to have `(the site | this site | here)` – in my testing the speech-to-text produces "the site" about as often as "this site" (must be hard to tell the difference).

ianb commented on extension/services/searchServices.toml:
```diff
@@ -37,6 +37,7 @@ bangSearch = "az"
 
 [wikipedia]
 bangSearch = "wikipedia"
+baseUrl = "https://en.wikipedia.org/wiki/"
```
This could be `https://en.wikipedia.org/` – the homepage doesn't have `/wiki/` for instance

danielamormocea commented:
I added some logic for the images search (duck duck go and google), using the information in the query, hope it's alright.
Also, for the metadata.music I think there are more changes to be made and I don't know whether that should be covered in a separate issue or not.

ianb commented on extension/background/serviceList.js:
```diff
@@ -164,14 +164,29 @@ export class Service {
     const service = services[name];
     if (tab.url.startsWith(service.baseUrl)) {
-      return name;
+      if (service.necessaryQuery === undefined) {
```
OK, I think I understand the code here, but some comments or different names might be in order.

`necessaryQuery` is necessary _here_, but if I don't know what your intention is it's not so clear. I think it would be clearer if it was named using `baseUrl`, since it's an additional filter to what `baseUrl` is used for. Like `baseUrlQueryParameters`

ianb commented on extension/background/serviceList.js:
```diff
@@ -164,14 +164,29 @@ export class Service {
-      return name;
+      if (service.necessaryQuery === undefined) {
+        serviceName = name;
+        continue;
```
And then I might leave a comment here: even if this matches, we still want to check if there's another more specific search provider that matches with query strings

danielamormocea commented:
@ianb , sorry I made a mistake and I reset my branch to a wrong commit. I redid all the changes and added the changes that you requested. I also added google as a service in searchServices toml, because I noticed that when the active tab is google the query 'Search on this site' was not working. Now it is working also on google. Hope it's alright.

ianb commented:
Github is being grumpy about letting me merge this PR, so I've merged master and recreated the PR in #1735
